mesop/dataclass_utils/dataclass_utils.py

```diff
@@ -54,13 +75,29 @@ def dataclass_with_defaults(cls: Type[C]) -> Type[C]:
       elif get_origin(type_hint) == dict:
         setattr(cls, name, field(default_factory=dict))
       elif isinstance(type_hint, type):
-        setattr(
-          cls, name, field(default_factory=dataclass_with_defaults(type_hint))
-        )
-
+        if has_parent(type_hint):
+          # If this isn't a simple class (i.e. it inherits from another class)
+          # then we will preserve its semantics (not try to set default values
+          # because it's not a dataclass) and instantiate it with each new instance
+          # of a state class.
+          setattr(cls, name, field(default_factory=type_hint))
+        else:
+          # If it's a simple dataclass (i.e. does not inherit from another class)
+          # then we will try to set default values and then wrap it with a dataclass
+          # decorator (if necessary).
+          setattr(
```

Comment:
Why do we try to wrap it in a dataclass if the class is not a dataclass? Does this affect serialization/deserialization?

Reply:
we wrap it, so 1) you can serialize/deserialize it and 2) dataclasses are generally convenient b/c it auto-generates stuff like __init__() and __repr__() - see https://docs.python.org/3/library/dataclasses.html